### ISSUE TITLE
PyTorch training

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#1486]
Using distributed training (`torch_distributed = {"param_sync_step": 100, "reduce_type": "param"}`), I got this cuFFT error (`RuntimeError: cuFFT error: CUFFT_INTERNAL_ERROR`). I assume that the aspect of distributed training is not relevant here.

Log:
```
RETURNN starting up, version 1.20240102.170719+git.83df39c7, date/time 2024-01-02-17-03-23 (UTC+0000), pid 3948138, cwd /work/asr4/zeyer/setups-data/comb
ined/2021-05-31/work/i6_core/returnn/training/ReturnnTrainingJob.ODn9wGvvwmOg/work, Python /work/tools/users/zeyer/py-envs/py3.11-torch2.1/bin/python3.11
RETURNN command line options: ['/u/zeyer/setups/combined/2021-05-31/work/i6_core/returnn/training/ReturnnTrainingJob.ODn9wGvvwmOg/output/returnn.config']
Hostname: cn-234
...
       name: NVIDIA GeForce GTX 1080 Ti
...
ep 1 train, step 900, ce 6.828, ctc_17 6.882, fer 0.939, mem_usage:cuda:3 6.8GB, 0.900 sec/step
ep 1 train, step 901, ce 6.917, ctc_17 6.959, fer 0.942, mem_usage:cuda:0 6.8GB, 0.545 sec/step
ep 1 train, step 901, ce 7.017, ctc_17 7.070, fer 0.958, mem_usage:cuda:3 6.8GB, 0.522 sec/step
ep 1 train, step 901, ce 6.629, ctc_17 6.653, fer 0.955, mem_usage:cuda:2 6.8GB, 0.565 sec/step
ep 1 train, step 901, ce 6.757, ctc_17 6.840, fer 0.943, mem_usage:cuda:1 6.8GB, 0.488 sec/step
ep 1 train, step 902, ce 7.114, ctc_17 7.167, fer 0.967, mem_usage:cuda:2 6.8GB, 0.504 sec/step
ep 1 train, step 902, ce 7.108, ctc_17 7.204, fer 0.942, mem_usage:cuda:3 6.8GB, 0.534 sec/step
ep 1 train, step 902, ce 6.895, ctc_17 6.974, fer 0.922, mem_usage:cuda:0 6.8GB, 0.573 sec/step
ep 1 train, step 902, ce 7.147, ctc_17 7.173, fer 0.945, mem_usage:cuda:1 6.8GB, 0.572 sec/step
ep 1 train, step 903, ce 6.867, ctc_17 6.915, fer 0.944, mem_usage:cuda:0 6.8GB, 0.435 sec/step
ep 1 train, step 903, ce 6.795, ctc_17 6.875, fer 0.933, mem_usage:cuda:3 6.8GB, 0.629 sec/step
ep 1 train, step 903, ce 6.628, ctc_17 6.557, fer 0.941, mem_usage:cuda:2 6.8GB, 0.666 sec/step
ep 1 train, step 903, ce 7.000, ctc_17 7.061, fer 0.955, mem_usage:cuda:1 6.8GB, 0.593 sec/step
Trained 904 steps, 0:09:12 elapsed (94.3% computing time)
Trained 904 steps, 0:09:12 elapsed (94.3% computing time)
Trained 904 steps, 0:09:12 elapsed (94.2% computing time)
Trained 904 steps, 0:09:12 elapsed (94.2% computing time)
Learning-rate-control: error key 'train_loss_ce' from {'train_loss_ce': 7.495721669989759, 'train_loss_fer': 0.9431181727904667, 'train_loss_ctc_17': 10.
243042048919284}
Learning-rate-control: error key 'train_loss_ce' from {'train_loss_ctc_17': 10.221398694536449, 'train_loss_fer': 0.943705191619096, 'train_loss_ce': 7.4
87026522100195}
Learning-rate-control: error key 'train_loss_ce' from {'train_loss_ce': 7.491586413589548, 'train_loss_ctc_17': 10.232071931271001, 'train_loss_fer': 0.9
425044664609388}
Learning-rate-control: error key 'train_loss_ce' from {'train_loss_fer': 0.9445714490897558, 'train_loss_ce': 7.4834289449724425, 'train_loss_ctc_17': 10
.282013659391682}
Total train loss: ce 7.496 ctc_17 10.243 fer 0.943
Total train loss: ce 7.492 ctc_17 10.232 fer 0.943
Total train loss: ce 7.483 ctc_17 10.282 fer 0.945
Memory usage (cuda:1): alloc cur 2.4GB alloc peak 6.8GB reserved cur 7.1GB reserved peak 7.1GB
Memory usage (cuda:3): alloc cur 2.4GB alloc peak 6.8GB reserved cur 7.1GB reserved peak 7.1GB
Memory usage (cuda:2): alloc cur 2.4GB alloc peak 6.8GB reserved cur 7.5GB reserved peak 7.5GB
Memory usage (cuda:3): alloc cur 2.4GB alloc peak 2.4GB reserved cur 7.1GB reserved peak 7.1GB
Memory usage (cuda:1): alloc cur 2.4GB alloc peak 2.4GB reserved cur 7.1GB reserved peak 7.1GB
Memory usage (cuda:2): alloc cur 2.4GB alloc peak 2.4GB reserved cur 7.5GB reserved peak 7.5GB
Total train loss: ce 7.487 ctc_17 10.221 fer 0.944
Memory usage (cuda:0): alloc cur 2.4GB alloc peak 6.8GB reserved cur 7.0GB reserved peak 7.0GB
Save model under /u/zeyer/setups/combined/2021-05-31/work/i6_core/returnn/training/ReturnnTrainingJob.ODn9wGvvwmOg/output/models/epoch.001.pt
Save optimizer under /u/zeyer/setups/combined/2021-05-31/work/i6_core/returnn/training/ReturnnTrainingJob.ODn9wGvvwmOg/output/models/epoch.001.opt.pt
Memory usage (cuda:0): alloc cur 2.2GB alloc peak 2.2GB reserved cur 7.0GB reserved peak 7.0GB
Evaluating dataset 'dev'
ep 1 dev eval, step 0, ce 7.086, ctc_17 7.055, fer 0.954, mem_usage:cuda:0 2.5GB
[2024-01-02 17:19:22,464] INFO: Run time: 0:16:07 CPU: 0.60% RSS: 61.72GB VMS: 1.15TB
ep 1 dev eval, step 1, ce 7.378, ctc_17 7.374, fer 0.964, mem_usage:cuda:0 2.5GB
...
ep 1 devtrain eval, step 393, ce 6.985, ctc_17 6.885, fer 0.972, mem_usage:cuda:0 2.8GB
ep 1 devtrain eval, step 394, ce 6.773, ctc_17 6.699, fer 0.944, mem_usage:cuda:0 2.8GB
ep 1 devtrain eval, step 395, ce 6.844, ctc_17 6.814, fer 0.941, mem_usage:cuda:0 2.8GB
ep 1 devtrain eval, step 396, ce 6.873, ctc_17 6.815, fer 0.955, mem_usage:cuda:0 2.8GB
ep 1 devtrain eval, step 397, ce 6.722, ctc_17 6.625, fer 0.954, mem_usage:cuda:0 2.8GB
ep 1 devtrain eval, step 398, ce 6.526, ctc_17 6.445, fer 0.917, mem_usage:cuda:0 2.8GB
ep 1 devtrain eval, step 399, ce 6.655, ctc_17 6.486, fer 0.955, mem_usage:cuda:0 2.8GB
Learning-rate-control: error key 'dev_loss_ce' from {'devtrain_loss_ctc_17': 6.960430699695994, 'devtrain_loss_fer': 0.9429784187378542, 'devtrain_loss_ce': 6.968572592061434}
dev: score ce 6.952 ctc_17 6.904 error 0.947 devtrain: score ce 6.969 ctc_17 6.960 error 0.943
Memory usage (cuda:0): alloc cur 2.5GB alloc peak 2.8GB reserved cur 7.0GB reserved peak 7.0GB
Only 1 epochs stored so far and keeping last 5 epochs and best 4 epochs, thus not cleaning up any epochs yet.
start epoch 2 global train step 904 with effective learning rate 1.2022460496613997e-05 ...
Memory usage (cuda:2): alloc cur 2.4GB alloc peak 2.4GB reserved cur 7.5GB reserved peak 7.5GB 
Memory usage (cuda:3): alloc cur 2.4GB alloc peak 2.4GB reserved cur 7.1GB reserved peak 7.1GB
Memory usage (cuda:0): alloc cur 2.5GB alloc peak 2.5GB reserved cur 7.0GB reserved peak 7.0GB
Memory usage (cuda:1): alloc cur 2.4GB alloc peak 2.4GB reserved cur 7.1GB reserved peak 7.1GB
ep 2 train, step 0, ce 6.977, ctc_17 7.016, fer 0.930, mem_usage:cuda:3 6.3GB, 0.713 sec/step
ep 2 train, step 0, ce 6.883, ctc_17 6.943, fer 0.977, mem_usage:cuda:0 6.4GB, 0.855 sec/step
ep 2 train, step 0, ce 6.910, ctc_17 6.931, fer 0.937, mem_usage:cuda:2 6.3GB, 0.793 sec/step
Unhandled exception <class 'RuntimeError'> in thread <_MainThread(MainThread, started 140347209949184)>, proc 3948138.
...
  File "/u/zeyer/setups/combined/2021-05-31/tools/returnn/returnn/torch/frontend/_backend.py", line 1845, in TorchBackend.stft
    line: y_raw = torch.stft(
              x_raw,
              n_fft=fft_length,
              hop_length=frame_step,
              # PyTorch anyway uses a window with size = n_fft internally.
              # So we just explicitly handle the window logic.
              win_length=fft_length,
              window=window_pt,
              center=False,
              return_complex=True,
          )
    locals:
      y_raw = <not found>
      torch = <global> <module 'torch' from '/work/tools/users/zeyer/py-envs/py3.11-torch2.1/lib/python3.11/site-packages/torch/__init__.py'>
      torch.stft = <global> <function stft at 0x7fa44d629120>
      x_raw = <local> tensor[20, 74912] n=1498240 (5.7Mb) x∈[2.016e+03, 275.000] μ=2.016e+03 σ=275.000 cuda:1
      n_fft = <not found>
      fft_length = <local> 512
      hop_length = <not found>
      frame_step = <local> 160
  File "/work/tools/users/zeyer/py-envs/py3.11-torch2.1/lib/python3.11/site-packages/torch/functional.py", line 650, in stft
    line: return _VF.stft(input, n_fft, hop_length, win_length, window,  # type: ignore[attr-defined]
                          normalized, onesided, return_complex)
    locals:
      _VF = <global> <module 'torch._VF'>
      _VF.stft = <global> <built-in method stft of type object at 0x7fa4b95aeaa0>
      input = <local> tensor[20, 74912] n=1498240 (5.7Mb) x∈[2.016e+03, 275.000] μ=2.016e+03 σ=275.000 cuda:1
      n_fft = <local> 512
      hop_length = <local> 160
      win_length = <local> 512
      window = <local> tensor[512] 2Kb x∈[2.016e+03, 275.000] μ=2.016e+03 σ=275.000 cuda:1
RuntimeError: cuFFT error: CUFFT_INTERNAL_ERROR
```

Now the process hangs.

**Comment by albertz** [comment]
```
zeyer@cn-234 ~ % nvidia-smi
Unable to determine the device handle for GPU0000:03:00.0: Unknown Error
```
And `dmesg`:
```
...
[  +0.000316] pcieport 0000:00:03.0: AER: Corrected error received: 0000:00:03.0
[  +0.000005] pcieport 0000:00:03.0: PCIe Bus Error: severity=Corrected, type=Data Link Layer, (Transmitter ID)
[  +0.000003] pcieport 0000:00:03.0:   device [8086:6f08] error status/mask=00001000/00002000
[  +0.000004] pcieport 0000:00:03.0:    [12] Timeout               
[  +0.000213] pcieport 0000:00:03.0: AER: Corrected error received: 0000:00:03.0
[  +0.000006] pcieport 0000:00:03.0: PCIe Bus Error: severity=Corrected, type=Data Link Layer, (Transmitter ID)
[  +0.000003] pcieport 0000:00:03.0:   device [8086:6f08] error status/mask=00001000/00002000
[  +0.000003] pcieport 0000:00:03.0:    [12] Timeout               
[  +0.001161] pcieport 0000:00:03.0: AER: Corrected error received: 0000:00:03.0
[  +0.000012] pcieport 0000:00:03.0: PCIe Bus Error: severity=Corrected, type=Data Link Layer, (Transmitter ID)
[  +0.000005] pcieport 0000:00:03.0:   device [8086:6f08] error status/mask=00001000/00002000
[  +0.000003] pcieport 0000:00:03.0:    [12] Timeout               
[  +0.000864] pcieport 0000:00:03.0: AER: Corrected error received: 0000:00:03.0
[  +0.000011] pcieport 0000:00:03.0: PCIe Bus Error: severity=Corrected, type=Data Link Layer, (Transmitter ID)
[  +0.000003] pcieport 0000:00:03.0:   device [8086:6f08] error status/mask=00001000/00002000
[  +0.000004] pcieport 0000:00:03.0:    [12] Timeout               
[Jan 2 17:17] pcieport 0000:00:03.0: AER: Uncorrected (Fatal) error received: 0000:00:03.0
[  +0.000018] pcieport 0000:00:03.0: PCIe Bus Error: severity=Uncorrected (Fatal), type=Transaction Layer, (Receiver ID)
[  +0.003292] pcieport 0000:00:03.0:   device [8086:6f08] error status/mask=00000020/00000000
[  +0.003017] pcieport 0000:00:03.0:    [ 5] SDES                   (First)
[  +0.003300] nvidia 0000:03:00.0: AER: can't recover (no error_detected callback)
[  +0.000009] snd_hda_intel 0000:03:00.1: AER: can't recover (no error_detected callback)
[  +0.315207] NVRM: Xid (PCI:0000:03:00): 79, pid='<unknown>', name=<unknown>, GPU has fallen off the bus.
[  +0.000015] NVRM: GPU 0000:03:00.0: GPU has fallen off the bus.
[  +0.000018] NVRM: A GPU crash dump has been created. If possible, please run
              NVRM: nvidia-bug-report.sh as root to collect this data before
              NVRM: the NVIDIA kernel module is unloaded.
[  +0.735941] pcieport 0000:00:03.0: AER: Root Port link has been reset (0)
[  +0.000047] pcieport 0000:00:03.0: AER: device recovery failed
```
So I guess this is a hardware issue.

**Comment by albertz** [comment]
Actually probably this here is a duplicate of #1465.
